### ISSUE TITLE
What is the prepared state doing?

### ISSUE THREAD

**Issue by alexanderrichards** [ganga-devs/ganga#154]
Looking at the hashes from the prepared state we realised we don't know how it works. Some concerns are:
1. If the hash is made from the Job `__repr__` then it would change whenever the state changes.
2. If the hash is made from the Application `__repr__` then nothing prevents me making a second prepared app with the same hash and hence we have a collision and an identity issue when choosing which prepared state to take from disk

If the intention of the prepared state hash is solely as a unique identifier to pick up the correct disk resident files then the python uuid package should probably be used instead as below:

``` python
import uuid
ident = uuid.uuid4()
```

If however the point is as above but includes the ability to detect if anything has changed then the above method wont work.

**However**: Off the top of my head I can't imagine why we would be tracking if a prepared state for an already prepared application had been changed. Surely they would just unprepare and reprepare thus creating a new uuid and the old files would be deleted when the counter for the old prepared state uuid hit zero.

In short we (or at least I :wink:) have no idea what's going on

**Comment by alexanderrichards** [comment]
So @rob-c has some additional info as he has just looked into it, he says and I quote:

>The prepared state folder 'config-xxxxx' is made using uuid and is stored in the job using the ShareDir class in ./Ganga/GPIDev/Lib/File/File.py
>
>The hash of a prepared application is calculated using the repr of the application using: calc_hash in ./Ganga/GPIDev/Adapters/IPrepareApp.py which makes use of just the application object which inherits from IPrepareApp.
>
>The hash is used in the GangaRepositoryXML (check_app_hash) in a step I'm keen to remove as it's performing an additional check on a read-only attribute which will (should) be locked from the users by being a read-only attribute for prepared apps.

**Comment by alexanderrichards** [comment]
so this looks at least then like the uuid is being used correctly. We should understand however what this application hash (which is used in GangaRepositoryXML.check_app_hash) is actually being used for and whether it makes sense or if indeed we can dump it which would be nice

**Comment by rob-c** [comment]
As far as I can tell from grepping the codebase the only place ```calc_hash``` is used is for providing an additional read-only check on the repr for the application. I'm keen to remove the check but if we remove ```calc_hash``` from the Schema we may trigger an un-prepare or worse for previously prepared jobs

**Comment by alexanderrichards** [comment]
Do jobs un-prepare themselves? I though the only time this was not a human initiated action was if the `unprepare_on_copy` (or whatever it's called) flag is set true and the user then copies a job.

**Comment by egede** [comment]
> Do jobs un-prepare themselves? I though the only time this was not a human initiated action was if 
> the unprepare_on_copy (or whatever it's called) flag is set true and the user then copies a job.

My understanding is also that applications inside jobs are only marked unprepared by a call to unprepare() or if the user sets the options that makes Ganga do this when copied. But I do not think that there is any way that Ganga will force an unprepare.

**Comment by alexanderrichards** [comment]
yes that was my understanding as well so I don't think that there should be a danger of triggering a mass un-preparing of user jobs.

**Comment by rob-c** [comment]
OK, the app isn't prepared if the hash changes. (or is removed?) when they go to an older version and we've changed it on disk.
At the very least there are a LOT of warnings which shouldn't be exposed to the user imo.
(several lines per write operation in the XML!)

I don't think we trigger a mas unprepare any more as there is no cleanup called on the application prepared repo.
